Refactor GetConnections to improve performance

hakonsbm opened the pull request:
This PR refactors parts of the `GetConnections` function to improve performance when specifying target nodes.

In the following benchmarks there are two populations of 5000 neurons each, connected with `all_to_all`. `GetConnections` is then called specifying target neurons, i.e. calling
```
<< /target target_nrns >> GetConnections
```
Times are given for the call to `GetConnections` only, in seconds.

| threads | v2.14 | ad5ef5c | refactor | v2.14 / refactor | ad5ef5c / refactor | 
|---------|---------|-------------|-----------------|-------------|-------------|
| 1       | 319.14  | 843.65      | 35.03           | 9.11        | 24.08       |
| 2       | 136.6   | 352.71      | 22.48           | 6.08        | 15.69       |
| 4       | 60.85   | 164         | 16.38           | 3.71        | 10.01       |
| 8       | 31.37   | 118.51      | 13.39           | 2.34        | 8.85        |
| 16      | 31.98   | 117.52      | 16.25           | 1.97        | 7.23        |

With the same setup, but now specifying both source and target neurons, i.e. calling
```
<< /source source_nrns /target target_nrns >> GetConnections
```

| threads | v2.14 | ad5ef5c | refactor | v2.14 / refactor | ad5ef5c / refactor |
|---------|---------|-------------|-----------------|-------------|-------------|
| 1       | 326.34  | 665.25      | 43.28           |  7.54       | 15.37       |
| 2       | 137.24  | 347.76      | 39.99           |  3.43       | 8.70        |
| 4       | 61.46   | 254.3       | 96.32           |  0.64       | 2.64        |
| 8       | 32.42   | 103.9       | 19.07           |  1.70       | 5.45        |
| 16      | 34.38   | 127.29      | 21.6            |  1.59       | 5.89        |

In both setups the refactored `GetConnections` is in almost all cases many times faster compared to both master (ad5ef5c) and v2.14. However with 4 threads the time for the refactored `GetConnections` is surprisingly long, but still an improvement on ad5ef5c.

Fixes #1096. Fixes #1085.

hakonsbm commented:
@stinebuu It does! Can you create a PR to my branch with the regressiontest?

alberto-antonietti commented:
@hakonsbm Thanks for taking care of this.

I tried your fix with previous (`ad5ef5c`) and current `master` (`201e722`) and I obtained the same improvements. I also got the same unexpectedly higher time.

However, I have seen that, even with your fix, the times needed by GetConnections changes with different connectivity rules used to generate the connections. With `v14 ` the time needed was more similar.

For example, creating 1.3M synapses between neurons A and B with a single thread:

```
A=nest.Create('iaf_cond_exp',650000)
B=nest.Create('iaf_cond_exp',2)

```
if connected with **all_to_all**:
`nest.Connect(A,B,'all_to_all', 'static_synapse')`
GetConnections takes **249s** (!!) with `201e722+PR` and 0.99s with `v2.14`

if connected with **fixed_indegree**:
`nest.Connect(A,B,{'rule': 'fixed_indegree','indegree':650000, 'multapses': False, 'autapses': False}, 'static_synapse')
`
GetConnections takes 2s with `201e722+PR` and 1.99s with `v2.14`

if connected with **fixed_outdegree**:
`nest.Connect(A,B,{'rule': 'fixed_outdegree','outdegree':2, 'multapses': False, 'autapses': False}, 'static_synapse')
`
GetConnections takes 1.7s with `201e722+PR` and 0.99s with `v2.14`

Are you experiencing the same?
Do you have any hint about the reason why exactly the same connectivity behaves differently?

hakonsbm commented:
@alberto-antonietti `GetConnections` after connecting all to all is slower for me as well, by a factor of ~10. With 5G (v2.16) connections have to be sorted when getting connections. Sorting the connections after connecting all to all takes longer than after connecting with fixed out- or indegree, probably because using all to all leaves the connection list in a state which takes longer to sort than using fixed out- or indegree.

heplesser commented:
@hakonsbm So is the factor 10 entirely down to the sorting? Since we are using our own implementation of quicksort, maybe we can figure out why things go wrong? I assume it may have something to do with quicksort's dislike of sorted data.

hakonsbm commented:
@heplesser The large increase in runtime for the all to all case is solely because of the sorting. Simulating for a short period before calling `GetConnections`, `GetConnections` takes about as long as in v2.14, as the connections are already sorted from simulating.

The difference between the fixed indegree case and the all to all case when sorting connections is number of swaps performed, mainly here

https://github.com/nest/nest-simulator/blob/386fb6ed77615a197d6c16fa39c738d10804a484/libnestutil/sort.h#L142-L150

Debugging with Valgrind, connecting 100000 -> 2 neurons with fixed indegree, sorting swaps elements 237277 times. Meanwhile connecting with all to all, elements are swapped 5075239 times.

Connections are sorted by the GIDs of the presynaptic neurons, and the initial order of connections is of course based on the connection rule. Fixed outdegree iterates source neurons, so connections are already sorted. Fixed indegree iterates target neurons and picks random source neurons, so connections are more or less random when sorting.

All to all iterates target neurons, then for each target neuron iterates source neurons (at least in this benchmark setup). This means that sources of the connections are the sequence of source neurons, repeated for each target neuron, e.g. [1, 2, ..., N, 1, 2, ..., N]  for N sources and 2 targets. This form is apparently a bad case for the sorting algorithm.

#### Is using std::sort a better alternative?
I benchmarked the NEST quicksort implementation versus `std::sort`, connecting 1M -> 2 neurons for fixed in- and outdegree, and 200000 -> 2 neurons for all to all, with the following results (all in seconds)

| Connection rule | NEST quicksort | `std::sort` |
|-----------------|----------------|-----------|
| Fixed outdegree | 0.0394041      | 0.0546829 |
| Fixed indegree  | 0.24474        | 0.187029  |
| All to all      | 5.82502        | 15.9669   |

There doesn't seem to be much to gain by using `std::sort` in these cases, especially as there would be additional overhead from adaptions to apply operations from `std::sort` on a second vector (which the NEST quicksort does natively). However, when connecting 30000 -> 3000 neurons all to all:

| Connection rule | NEST quicksort | `std::sort` |
|-----------------|----------------|-----------|
| All to all      | 55.5934        | 4.47839   |

Here the NEST quicksort implementation struggles, while `std::sort` is more than 12 times as fast.

jakobj commented:
have you tried it with 30000 -> 3000 +- a few? the issue seems to be how the pivot element is chosen in an array with regular structure. for a badly chosen pivot element, quicksort needs an unnecessary high number of swapping operations as it unevenly divides the elements. usually one would protect against these kinds of performance issues by first shuffling the original array (complexity linear in N) which we have failed to do here. lesson learned. seems like we have three options:
1. implement a shuffle operation which keeps sources and connections in the correct relative order and apply it before sorting the array
2. find a smarter way to choose the pivot
3. find a library implementation of quicksort that is already battle tested and can do what we need to replace this implementation

hakonsbm commented:
@jakobj I have implemented an alternative sort method using Boost sort. Their algorithm avoids this pivot problem, and is quite a lot faster when sorting random numbers as well. There will be a PR soon.

jougs commented:
@jakobj: would you be willing to serve as reviewer here? I have the feeling you are more into this than I am... Thanks!

jakobj commented on nestkernel/connection_manager.cpp:
```diff
@@ -1221,23 +1221,23 @@ nest::ConnectionManager::get_connections(
+            synapse_label,
+            conns_in_thread );
+        }
+
```
as far as I can tell this loop uses at most `n_conns * n_targets` comparisons (loop + linear complexity of `std::find` in `get_connection_with_specified_targets`). the previous implementation also should use the same amount of comparisons in the worst case. your new implementation is much faster, so where did my reasoning fail?

hakonsbm replied:
In the previous implementation both the vector of specified targets and the vector of local connections are always completely iterated, so there are always `n_conns * n_targets` comparisons, like you say. This implementation iterates over the connections first, so it only has to check if the target of the connection is among the specified targets, and therefore doesn't have to check all of them for every connection. Also `std::find` may do some hardware-near optimizations.

jakobj replied:
ok, I think I understand. while the worst case runtime is identical for both implementations, the average case is much better in the new one. :+1:

jakobj commented on testsuite/regressiontests/issue-1085.sli:
```diff
@@ -0,0 +1,87 @@
+/*
```
should this be part of this PR?

hakonsbm replied:
Yes. @stinebuu suggested as both this PR and #1093 fixes #1085, #1093 should be closed in favor of this. Therefore this PR got the regressiontest.

jakobj replied:
ok

jakobj commented:
looks good! thanks for your effort :+1:
